### PR TITLE
Return an error from `mbedtls_cipher_set_iv` for an invalid IV length with ChaCha20 and ChaCha20+Poly

### PR DESCRIPTION
**Pull request by AndrzejKurek**
`mbedtls_cipher_set_iv` was silently overwritting the `iv_len` set by the caller, thus leading to potential API misuse.
Fixes https://github.com/ARMmbed/mbedtls/issues/4301.

### CODE REVIEW

**Review by mprse — CHANGES_REQUESTED** [review]
All looks good. I left few minor comments.

File: tests/suites/test_suite_cipher.chachapoly.data [hunk 1]
```diff
@@ -121,3 +121,23 @@ auth_crypt_tv:MBEDTLS_CIPHER_CHACHA20_POLY1305:"1c9240a5eb55d38af333888604f6b5f0
...
+depends_on:MBEDTLS_CHACHAPOLY_C
+check_iv:MBEDTLS_CIPHER_CHACHA20_POLY1305:"CHACHA20-POLY1305":0:MBEDTLS_ERR_CIPHER_BAD_INPUT_DATA
+
+ChaCha20+Poly1305 IV Length 11
+depends_on:MBEDTLS_CHACHAPOLY_C
+check_iv:MBEDTLS_CIPHER_CHACHA20_POLY1305:"CHACHA20-POLY1305":11:MBEDTLS_ERR_CIPHER_BAD_INPUT_DATA
+
+ChaCha20+Poly1305 IV Length 12
+depends_on:MBEDTLS_CHACHAPOLY_C
+check_iv:MBEDTLS_CIPHER_CHACHA20_POLY1305:"CHACHA20-POLY1305":12:0
```
[the remark this answers is not in the document]
Comment (AndrzejKurek) [reply tests/suites/test_suite_cipher.chachapoly.data:135]: There's no macro here, citing `cipher.h`:

>  * \return          \c 0 on success.
>  * \return          #MBEDTLS_ERR_CIPHER_BAD_INPUT_DATA on
>  *                  parameter-verification failure.
>  */
> int mbedtls_cipher_set_iv( mbedtls_cipher_context_t *ctx,
>                            const unsigned char *iv,
>                            size_t iv_len );

File: tests/suites/test_suite_cipher.function [hunk 2]
```diff
@@ -551,6 +555,11 @@ void dec_empty_buf( int cipher,
     /* Initialise context */
     cipher_info = mbedtls_cipher_info_from_type( cipher );
     TEST_ASSERT( NULL != cipher_info);
+
+    if( cipher_info->type == MBEDTLS_CIPHER_CHACHA20 ||
+        cipher_info->type == MBEDTLS_CIPHER_CHACHA20_POLY1305 )
+        iv_len = 12;
```
[the remark this answers is not in the document]
Comment (gabor-mezei-arm) [reply tests/suites/test_suite_cipher.function:599]: There is no need to dynamically allocate memory for the iv buffer.

**Review by gabor-mezei-arm — CHANGES_REQUESTED** [review]

File: tests/suites/test_suite_cipher.function [hunk 3]
```diff
@@ -589,6 +593,14 @@ void dec_empty_buf( int cipher,
...
+
+    ASSERT_ALLOC( iv, iv_len );
```
Comment (gabor-mezei-arm) [inline tests/suites/test_suite_cipher.function:601]: There is no need to dynamically allocate memory for the iv buffer.

Comment (AndrzejKurek) [reply tests/suites/test_suite_cipher.function:601]: There's one advantage of only allocating the minimal appropriate size: if the `iv_len` is overwritten by anything (as was the case in the original code), passing a smaller buffer will alert any static analysis tools, as the code might want to access data beyond it.

File: tests/suites/test_suite_cipher.function (hunk 2 above)
[the remark this answers is not in the document]
Comment (AndrzejKurek) [reply tests/suites/test_suite_cipher.function:599]: Done that :) It has an additional advantage as described in my comment below.

File: tests/suites/test_suite_cipher.chacha20.data [hunk 4]
```diff
@@ -1,3 +1,7 @@
+Decrypt empty buffer
+depends_on:MBEDTLS_CHACHAPOLY_C
```
[the remark this answers is not in the document]
Comment (AndrzejKurek) [reply tests/suites/test_suite_cipher.chacha20.data]: Fixed, thanks!

**Review by mprse — APPROVED** [review]
LGTM

**Review by gabor-mezei-arm — APPROVED** [review]
Looks good to me

**Pull request merged**